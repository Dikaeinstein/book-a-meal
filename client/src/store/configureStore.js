import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { createLogger } from 'redux-logger';
import throttle from 'lodash.throttle';
import { loadState, saveState } from '../helpers/localStorage';
import initialState from '../reducers/initialState';
import rootReducer from '../reducers';

/**
 * Return redux middleware
 */
const getMiddleware = () => {
  const logger = createLogger();

  const middleware =
    process.env.NODE_ENV !== 'production'
    && process.env.NODE_ENV !== 'test'
      ? applyMiddleware(thunk, logger)
      : applyMiddleware(thunk);

  return middleware;
};

/**
 * Configure the redux store
 *
 * @returns {object} Redux store
 */
const configureStore = () => {
  const persistedState = loadState();

  const state = persistedState
    ? { ...initialState, ...persistedState }
    : initialState;

  const store = createStore(
    rootReducer,
    state,
    compose(
      getMiddleware(),
      window.devToolsExtension ? window.devToolsExtension() : empty => empty,
    ),
  );

  // Throttle the saveState function to be called once every second
  store.subscribe(throttle(() => {
    saveState({ orders: store.getState().orders });
  }, 1000));

  return store;
};

export default configureStore;
