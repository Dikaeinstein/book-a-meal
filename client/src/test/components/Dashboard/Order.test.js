import React from 'react';
import { shallow } from 'enzyme';
import sweetalert from 'sweetalert';
import configureStore from 'redux-mock-store';
import ConnectedOrder, { Order } from '../../../components/Dashboard/Order';

jest.mock('sweetalert');

describe('<Order />', () => {
  const order = {
    id: 11,
    quantity: 1,
    total: '4000',
    status: 'delivered',
    expired: false,
    mealId: 3,
    userId: 4,
    meal: {
      id: 3,
      name: 'Amala and Ewedu',
      price: '4000',
    },
    user: {
      id: 4,
      name: 'Chinonso Okorie',
    },
  };
  it('should render with correct props', () => {
    const mockStore = configureStore();
    const store = mockStore({ dashboard: { isUpdating: false } });
    const wrapper = shallow(<ConnectedOrder
      store={store}
      order={order}
    />);
    expect(wrapper.props().isUpdating).toBeFalsy();
  });
  it('should call handleCancel when order is cancelled', () => {
    const modifyOrder = jest.fn();
    const wrapper = shallow(<Order
      modifyOrder={modifyOrder}
      order={order}
    />);
    const cancelOrder = wrapper.find('.action-btn').last();
    sweetalert.mockResolvedValue(Promise.resolve(true));
    cancelOrder.simulate('click');
    expect(wrapper).toMatchSnapshot();
    sweetalert.mockResolvedValue(Promise.reject(new Error()));
    cancelOrder.simulate('click');
    expect(wrapper).toMatchSnapshot();
    sweetalert.mockResolvedValue(Promise.reject());
    cancelOrder.simulate('click');
    expect(wrapper).toMatchSnapshot();
  });
  it('should call handleDeliver when order is delivered', () => {
    const modifyOrder = jest.fn();
    const wrapper = shallow(<Order
      modifyOrder={modifyOrder}
      order={order}
    />);
    const deliverOrder = wrapper.find('.action-btn').first();
    sweetalert.mockResolvedValue(Promise.resolve(true));
    deliverOrder.simulate('click');
    expect(wrapper).toMatchSnapshot();
    sweetalert.mockResolvedValue(Promise.reject(new Error()));
    deliverOrder.simulate('click');
    expect(wrapper).toMatchSnapshot();
    sweetalert.mockResolvedValue(Promise.reject());
    deliverOrder.simulate('click');
    expect(wrapper).toMatchSnapshot();
  });
  it('should match snapshot', () => {
    let tree = shallow(<Order order={order} modifyOrder={jest.fn()} />);
    expect(tree).toMatchSnapshot();
    tree = shallow(<Order
      order={{ ...order, status: 'cancelled' }}
      modifyOrder={jest.fn()}
    />);
    expect(tree).toMatchSnapshot();
    tree = shallow(<Order
      order={{ ...order, status: 'pending' }}
      modifyOrder={jest.fn()}
    />);
    expect(tree).toMatchSnapshot();
    tree = shallow(<Order
      order={{
        ...order,
        user: { name: '' },
        meal: { name: '', price: '' },
      }}
      modifyOrder={jest.fn()}
    />);
    expect(tree).toMatchSnapshot();
  });
});
