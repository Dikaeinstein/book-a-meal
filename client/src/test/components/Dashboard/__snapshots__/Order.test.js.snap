// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Order /> should call handleCancel when order is cancelled 1`] = `
<tr
  key="11"
>
  <td>
    11
  </td>
  <td>
    Amala and Ewedu
  </td>
  <td>
    Chinonso Okorie
  </td>
  <td>
    4000
  </td>
  <td>
    1
  </td>
  <td>
    4000
  </td>
  <td
    className="text-success"
  >
    delivered
  </td>
  <td>
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <button
        className="action-btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#ccc",
            "cursor": "default",
          }
        }
        title="Deliver order"
        value="Deliver"
      >
        Deliver
      </button>
      <button
        className="action-btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#ccc",
            "cursor": "default",
          }
        }
        title="Cancel order"
        value="Cancel"
      >
        Cancel
      </button>
    </div>
  </td>
</tr>
`;

exports[`<Order /> should call handleCancel when order is cancelled 2`] = `
<tr
  key="11"
>
  <td>
    11
  </td>
  <td>
    Amala and Ewedu
  </td>
  <td>
    Chinonso Okorie
  </td>
  <td>
    4000
  </td>
  <td>
    1
  </td>
  <td>
    4000
  </td>
  <td
    className="text-success"
  >
    delivered
  </td>
  <td>
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <button
        className="action-btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#ccc",
            "cursor": "default",
          }
        }
        title="Deliver order"
        value="Deliver"
      >
        Deliver
      </button>
      <button
        className="action-btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#ccc",
            "cursor": "default",
          }
        }
        title="Cancel order"
        value="Cancel"
      >
        Cancel
      </button>
    </div>
  </td>
</tr>
`;

exports[`<Order /> should call handleCancel when order is cancelled 3`] = `
<tr
  key="11"
>
  <td>
    11
  </td>
  <td>
    Amala and Ewedu
  </td>
  <td>
    Chinonso Okorie
  </td>
  <td>
    4000
  </td>
  <td>
    1
  </td>
  <td>
    4000
  </td>
  <td
    className="text-success"
  >
    delivered
  </td>
  <td>
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <button
        className="action-btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#ccc",
            "cursor": "default",
          }
        }
        title="Deliver order"
        value="Deliver"
      >
        Deliver
      </button>
      <button
        className="action-btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#ccc",
            "cursor": "default",
          }
        }
        title="Cancel order"
        value="Cancel"
      >
        Cancel
      </button>
    </div>
  </td>
</tr>
`;

exports[`<Order /> should call handleDeliver when order is delivered 1`] = `
<tr
  key="11"
>
  <td>
    11
  </td>
  <td>
    Amala and Ewedu
  </td>
  <td>
    Chinonso Okorie
  </td>
  <td>
    4000
  </td>
  <td>
    1
  </td>
  <td>
    4000
  </td>
  <td
    className="text-success"
  >
    delivered
  </td>
  <td>
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <button
        className="action-btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#ccc",
            "cursor": "default",
          }
        }
        title="Deliver order"
        value="Deliver"
      >
        Deliver
      </button>
      <button
        className="action-btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#ccc",
            "cursor": "default",
          }
        }
        title="Cancel order"
        value="Cancel"
      >
        Cancel
      </button>
    </div>
  </td>
</tr>
`;

exports[`<Order /> should call handleDeliver when order is delivered 2`] = `
<tr
  key="11"
>
  <td>
    11
  </td>
  <td>
    Amala and Ewedu
  </td>
  <td>
    Chinonso Okorie
  </td>
  <td>
    4000
  </td>
  <td>
    1
  </td>
  <td>
    4000
  </td>
  <td
    className="text-success"
  >
    delivered
  </td>
  <td>
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <button
        className="action-btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#ccc",
            "cursor": "default",
          }
        }
        title="Deliver order"
        value="Deliver"
      >
        Deliver
      </button>
      <button
        className="action-btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#ccc",
            "cursor": "default",
          }
        }
        title="Cancel order"
        value="Cancel"
      >
        Cancel
      </button>
    </div>
  </td>
</tr>
`;

exports[`<Order /> should call handleDeliver when order is delivered 3`] = `
<tr
  key="11"
>
  <td>
    11
  </td>
  <td>
    Amala and Ewedu
  </td>
  <td>
    Chinonso Okorie
  </td>
  <td>
    4000
  </td>
  <td>
    1
  </td>
  <td>
    4000
  </td>
  <td
    className="text-success"
  >
    delivered
  </td>
  <td>
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <button
        className="action-btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#ccc",
            "cursor": "default",
          }
        }
        title="Deliver order"
        value="Deliver"
      >
        Deliver
      </button>
      <button
        className="action-btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#ccc",
            "cursor": "default",
          }
        }
        title="Cancel order"
        value="Cancel"
      >
        Cancel
      </button>
    </div>
  </td>
</tr>
`;

exports[`<Order /> should match snapshot 1`] = `
<tr
  key="11"
>
  <td>
    11
  </td>
  <td>
    Amala and Ewedu
  </td>
  <td>
    Chinonso Okorie
  </td>
  <td>
    4000
  </td>
  <td>
    1
  </td>
  <td>
    4000
  </td>
  <td
    className="text-success"
  >
    delivered
  </td>
  <td>
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <button
        className="action-btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#ccc",
            "cursor": "default",
          }
        }
        title="Deliver order"
        value="Deliver"
      >
        Deliver
      </button>
      <button
        className="action-btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#ccc",
            "cursor": "default",
          }
        }
        title="Cancel order"
        value="Cancel"
      >
        Cancel
      </button>
    </div>
  </td>
</tr>
`;

exports[`<Order /> should match snapshot 2`] = `
<tr
  key="11"
>
  <td>
    11
  </td>
  <td>
    Amala and Ewedu
  </td>
  <td>
    Chinonso Okorie
  </td>
  <td>
    4000
  </td>
  <td>
    1
  </td>
  <td>
    4000
  </td>
  <td
    className="text-danger"
  >
    cancelled
  </td>
  <td>
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <button
        className="action-btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#ccc",
            "cursor": "default",
          }
        }
        title="Deliver order"
        value="Deliver"
      >
        Deliver
      </button>
      <button
        className="action-btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#ccc",
            "cursor": "default",
          }
        }
        title="Cancel order"
        value="Cancel"
      >
        Cancel
      </button>
    </div>
  </td>
</tr>
`;

exports[`<Order /> should match snapshot 3`] = `
<tr
  key="11"
>
  <td>
    11
  </td>
  <td>
    Amala and Ewedu
  </td>
  <td>
    Chinonso Okorie
  </td>
  <td>
    4000
  </td>
  <td>
    1
  </td>
  <td>
    4000
  </td>
  <td
    className="text-warning"
  >
    pending
  </td>
  <td>
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <button
        className="action-btn"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#28a745",
          }
        }
        title="Deliver order"
        value="Deliver"
      >
        Deliver
      </button>
      <button
        className="action-btn"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#dc3545",
          }
        }
        title="Cancel order"
        value="Cancel"
      >
        Cancel
      </button>
    </div>
  </td>
</tr>
`;

exports[`<Order /> should match snapshot 4`] = `
<tr
  key="11"
>
  <td>
    11
  </td>
  <td>
    -
  </td>
  <td>
    -
  </td>
  <td>
    -
  </td>
  <td>
    1
  </td>
  <td>
    4000
  </td>
  <td
    className="text-success"
  >
    delivered
  </td>
  <td>
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <button
        className="action-btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#ccc",
            "cursor": "default",
          }
        }
        title="Deliver order"
        value="Deliver"
      >
        Deliver
      </button>
      <button
        className="action-btn"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#ccc",
            "cursor": "default",
          }
        }
        title="Cancel order"
        value="Cancel"
      >
        Cancel
      </button>
    </div>
  </td>
</tr>
`;
